### PR TITLE
Add the new Results and ResultsInput models to Repo

### PR DESCRIPTION
**Pull request by Lnaden**
Outstanding issues:
* `basis_spec`'s are omitted for now to get this model in. @dgasmith indicates this is probably fine for now
* Missing tests. Do we have a good place to pull some sample data from that isn't me just writing the dicts that also does not have to execute any computation though something like QCEngine?

### CODE REVIEW

**Comment by Lnaden** [comment]
Follow up question: Do people see the need to add methods to these classes, or just leave them as effectively lookup models?

**Comment by loriab** [comment]
well, I do have a list (https://github.com/loriab/qcdb/blob/master/qcdb/modelchems.py#L452). but I think leave it free-form.

I'm already having fun with the required basis field for methods like hf-3c, gaussian-2, and sns-mp2. :-) `basis='(auto)'` coming in handy.

File: qcelemental/models/results.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    mp2_total_energy: float = None
+    calcinfo_nbasis: int = None
+    calcinfo_nmo: int = None
+    calcinfo_nalpha: int = None
+    calcinfo_nbeta: int = None
+    calcinfo_natom: int = None
+    return_energy: float = None
+
+    class Config:
+        allow_extra = False
```
Comment (dgasmith) [inline qcelemental/models/results.py]: `allow_extra = True` this really was best left as a dictionary field that was not validated at the moment.

File: qcelemental/models/results.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class ErrorEnum(str, Enum):
+    convergence_error = "convergence_error"
+    file_error = "file_error"
+    memory_error = "memory_error"
```
Comment (dgasmith) [inline qcelemental/models/results.py]: Not yet strict enum.

Comment (Lnaden) [reply qcelemental/models/results.py]: The schema makes it a strict enum at the moment, but we could convert it back to a `Dict[str, str]` for now

File: qcelemental/models/results.py (hunk 1 above)
Comment (dgasmith) [reply qcelemental/models/results.py]: Eh, I say leave it as is for the moment.

File: qcelemental/models/results.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    class Config:
+        allow_mutation = False
+
+
+class Results(ResultsInput):
+    properties: Properties
+    success: bool
+    error: Error = None
+    return_result: Union[float, List[float]]
```
Comment (dgasmith) [inline qcelemental/models/results.py]: This needs a `allow_extra=True`. That might overwrite the `Config`, if so the `allow_mutation=False` is good here.

File: qcelemental/models/results.py (hunk 2 above)
Comment (dgasmith) [reply qcelemental/models/results.py]: Please do.

File: qcelemental/models/results.py (hunk 3 above)
Comment (Lnaden) [reply qcelemental/models/results.py]: Just to confirm, you want the `Result` object to support additional arbitrary fields and still be immutable, correct?

**Comment by dgasmith** [comment]
Is this ready to go?

File: qcelemental/models/results.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    error_type: Dict[str, str]  # Error enumeration not yet strict
+    error_message: str
+
+    class Config:
+        allow_extra = False
+
+
+### Primary models
+
+class ResultsInput(BaseModel):
```
Comment (dgasmith) [inline qcelemental/models/results.py]: This should be `Result` singular.

Comment (Lnaden) [reply qcelemental/models/results.py]: Do you want `Result` singular as well?

**Comment by Lnaden** [comment]
> Is this ready to go?

Barring the plurality check, are there any test data you can point me at to add to the model here before we merge it? Otherwise, its probably ready

**Comment by Lnaden** [comment]
This should be good to go from my perspective. I think as we integrate this into the rest of QCArchive, we will have some decent test data we can add back in

**Pull request merged**